snowfall/objectives/mmi.py

```diff
@@ -32,20 +32,65 @@ def forward(
             texts: List,
             supervision_segments: torch.Tensor
     ) -> Tuple[torch.Tensor, int, int]:
-        num_graphs, den_graphs = self.graph_compiler.compile(texts, self.P)
+        num_graphs, den_graphs = self.graph_compiler.compile(
+            texts, self.P, replicate_den=False)
+
         dense_fsa_vec = k2.DenseFsaVec(nnet_output, supervision_segments)
 
-        num_lats = k2.intersect_dense(num_graphs, dense_fsa_vec, 10.0)
-        den_lats = k2.intersect_dense(den_graphs, dense_fsa_vec, 10.0)
+        device = num_graphs.device
+
+        num_fsas = num_graphs.shape[0]
+        assert dense_fsa_vec.dim0() == num_fsas
+
+        assert den_graphs.shape[0] == 1
+
+        # the aux_labels of num_graphs is k2.RaggedInt
+        # but it is torch.Tensor for den_graphs.
+        #
+        # The following converts den_graphs.aux_labels
+        # from torch.Tensor to k2.RaggedInt so that
+        # we can use k2.append() later
+        ragged_shape = k2.ragged.regular_ragged_shape(
+            dim0=den_graphs.aux_labels.numel(), dim1=1).to(device)
+
+        den_graphs.aux_labels = k2.RaggedInt(ragged_shape,
```

Comment:
I think it would be more elegant to make this a function invoked like `den_graphs.convert_attr_to_ragged('aux_labels', remove_eps=True)`.
I imagine this will still have [-1] on the final-arcs, which is probably what we want
as the num graphs likely have this.

Reply:
... Incidentally, the opposite to this function would, I guess, be convert_attr_to_linear(), which would have to call ExpandArcs().  However this would have to be a function in fsa_algo, as in: `x = k2.convert_attr_to_linear(x, 'aux_labels')`, as it might modify the FSA.  We can just keep this in mind as something we might want to do in the future.

Reply:
> I think it would be more elegant to make this a function invoked like `den_graphs.convert_attr_to_ragged('aux_labels', remove_eps=True)`.
> 
> I imagine this will still have [-1] on the final-arcs, which is probably what we want
> 
> as the num graphs likely have this.
> 
> 
> 
> 

ok, will move it to k2.

Reply:
> den_graphs.convert_attr_to_ragged('aux_labels', remove_eps=True).

I assume that `aux_labels` for this function is a 1-D tensor. How do we handle the case for 2-D or N-D tensors?
Why do we need the extra argument `remove_eps`?

Reply:
Currently we don't have to support such a thing for N-d tensors (eventually we may decide to allow ragged+regular arrays like that).
The remove_eps is because in general when we store labels as ragged arrays, we remove the zero elements.  
In this particular situation it doesn't matter because we won't actually be doing anything with these labels.